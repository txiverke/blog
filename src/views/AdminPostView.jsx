// @flow

import React from 'react'
import Helmet from 'react-helmet'
import { connect } from 'react-redux'
import ReactMessages from 'react-messages'

import Loader from '../components/Loader'
import AdminPostItem from '../components/AdminPostItem'
import { loadPostItem, updatePostData } from '../actions/postActionCreators'
import { isAuthenticated } from '../actions/userActionCreators'

class AdminPostView extends React.Component {
  state = {
    data: {}
  }

  props: {
    dispatch: Function,
    item: Item,
    authenticate: Auth,
    match: Object
  }

  componentDidMount() {
    const { dispatch, match } = this.props
    dispatch(loadPostItem(match.params.id))
    dispatch(isAuthenticated())
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ data: nextProps.item.data })
  }

  updatePost = (obj: Object, id: string) => {
    const { dispatch } = this.props
    dispatch(updatePostData(obj, id))
  }

  render() {
    const { message, error, completed } = this.props.item
    const { data } = this.state

    if (completed && Object.keys(data).length) {
      return (
        <section className="app-view pb5">
          <Helmet 
            title="Admin page" 
            meta={[
              { name: "description", content: "Admin page" },
              { property: "og:title", content: "Admin page" }
            ]}
          />  
          <ReactMessages message={message} error={error} next={true} />
          <AdminPostItem 
            label="Update post" 
            handlePost={(obj, id) => this.updatePost(obj, id)} 
            data={data}
          />
        </section>
      )
    }

    return (
      <div className="app-view app-view-centered">
        <Loader msg={message} />
      </div>
    )
  }
}

const mapStateToProps = state => ({ 
  authenticate: state.authenticate,
  item: state.item, 
})

export default connect(mapStateToProps)(AdminPostView)