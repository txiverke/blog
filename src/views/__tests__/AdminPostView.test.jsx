// flow

import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { Unwrapped as UnwrappedAdminPostView } from '../AdminPostView'

const properties = {
  dispatch: jest.fn(),
  item: {
    data: {},
    message: 'Just a fake message',
    error: false,
    completed: true
  },
  authenticate: { 
    completed: true,
    data: '',
    message: '',
    error: false
  },
  match: { params: { id: '1'}}
}

function setup() {
  const props = properties

  const enzymeWrapper = mount(
    <MemoryRouter initialEntries={['/test']}>
      <UnwrappedAdminPostView {...props} />
    </MemoryRouter>
  )

  return {
    props,
    enzymeWrapper
  }
}

describe('<AdminPostView />', () => {
  it('should render itself', () => {
    const component = shallow(<UnwrappedAdminPostView {...properties} />)
    expect(component).toMatchSnapshot()
  })

  it('should render Redirect if user is not authenticate', () => {
    properties.authenticate.data = ''
    const component = shallow(<UnwrappedAdminPostView {...properties} />)
    expect(component.find('Redirect').length).toEqual(1)
  })

  it('should render Loader if no data', () => {
    properties.authenticate.data = 'test'
    properties.item.completed = false
    const component = shallow(<UnwrappedAdminPostView {...properties} />)
    expect(component.find('Redirect').length).toEqual(0)
    expect(component.find('Loader').length).toEqual(1)

  })

  it('should render Post data if there is data', () => {
    properties.item.completed = true
    properties.item.data = { test: true }
    const component = shallow(<UnwrappedAdminPostView {...properties} />)
    expect(component.find('Redirect').length).toEqual(0)
    expect(component.find('Loader').length).toEqual(0)
    expect(component.find('AdminPostItem').length).toEqual(1)
  })

  it('should call updateProject and dispatch updateProjectData action', () => {
    const { enzymeWrapper, props } = setup()  
    const item = enzymeWrapper.find('AdminPostItem')
    const currentCalls = props.dispatch.mock.calls.length
    item.props().handlePost({}, '')
    expect(props.dispatch.mock.calls.length).toBe(currentCalls)
    item.props().handlePost({test: true}, '1')
    expect(props.dispatch.mock.calls.length).toBe(currentCalls + 1)
  })
})