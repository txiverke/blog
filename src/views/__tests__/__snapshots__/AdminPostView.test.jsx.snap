// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminPostView /> should render itself 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminPostView
    authenticate={Object {}}
    dispatch={[Function]}
    item={
        Object {
            "completed": true,
            "error": false,
            "message": "Just a fake message",
          }
    }
    match={
        Object {
            "params": Object {
              "id": "1",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Loader
        msg="Just a fake message"
/>,
      "className": "app-view app-view-centered",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "msg": "Just a fake message",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Loader
          msg="Just a fake message"
/>,
        "className": "app-view app-view-centered",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "msg": "Just a fake message",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
